File: src/gigaloom/tools/mcp/apps/bridge.py
"""Instance-bound JSON-RPC validation for the MCP App postMessage bridge."""

from __future__ import annotations

import hashlib
import hmac
import json
from threading import RLock
from typing import Callable, Mapping

from .contracts import (
    MCPAppBridgeRequest,
    MCPAppFrameBinding,
    MCPAppLimits,
)
from .errors import MCPAppChannelError
from .evidence import MCPAppSecurityEvent

MCP_APP_ALLOWED_METHODS = frozenset({"ui/ready", "ui/userChoice"})


class MCPAppBridge:
    """Validate one frame's messages and track replay/outstanding state."""

    def __init__(
        self,
        *,
        instance_id: str,
        source_id: str,
        channel_id: str,
        nonce: str,
        binding: MCPAppFrameBinding,
        limits: MCPAppLimits,
        record_event: Callable[[MCPAppSecurityEvent], None] | None = None,
    ) -> None:
        self.instance_id = instance_id
        self.source_id = source_id
        self.channel_id = channel_id
        self.nonce = nonce
        self.binding = binding
        self._limits = limits
        self._record_event = record_event
        self._outstanding: set[str | int] = set()
        self._seen: set[str | int] = set()
        self._closed = False
        self._lock = RLock()

    def accept(self, *, source_id: str, raw_message: bytes) -> MCPAppBridgeRequest:
        """Validate and register one app-to-host JSON-RPC request."""
        with self._lock:
            try:
                request = self._accept_locked(
                    source_id=source_id, raw_message=raw_message
                )
            except MCPAppChannelError as exc:
                self._audit(exc.code)
                raise
            self._audit(
                "request_accepted",
                method=request.method,
                request_id=request.request_id,
            )
            return request

    def complete(self, request_id: str | int) -> bool:
        """Complete an outstanding request; ignore audited late responses."""
        with self._lock:
            if self._closed or request_id not in self._outstanding:
                self._audit("late_response", request_id=request_id)
                return False
            self._outstanding.remove(request_id)
            self._audit("request_completed", request_id=request_id)
            return True

    def teardown(self) -> tuple[str | int, ...]:
        """Close the frame and cancel all pending request ids."""
        with self._lock:
            cancelled = tuple(sorted(self._outstanding, key=str))
            self._outstanding.clear()
            self._closed = True
            self._audit("frame_destroyed")
            return cancelled

    def _accept_locked(
        self,
        *,
        source_id: str,
        raw_message: bytes,
    ) -> MCPAppBridgeRequest:
        if self._closed:
            raise MCPAppChannelError("frame_closed", "MCP App frame is closed")
        if not hmac.compare_digest(source_id, self.source_id):
            raise MCPAppChannelError(
                "source_mismatch", "MCP App message source does not match its frame"
            )
        if not raw_message or len(raw_message) > self._limits.max_post_message_bytes:
            raise MCPAppChannelError(
                "payload_limit", "MCP App message exceeds the configured byte limit"
            )
        try:
            payload = json.loads(raw_message)
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise MCPAppChannelError(
                "invalid_json", "MCP App message is not valid JSON"
            ) from exc
        if not isinstance(payload, dict) or _json_depth(payload) > 32:
            raise MCPAppChannelError(
                "invalid_jsonrpc", "MCP App message must be a bounded JSON object"
            )
        allowed_keys = {"jsonrpc", "id", "method", "params", "channelId", "nonce"}
        if set(payload) - allowed_keys or payload.get("jsonrpc") != "2.0":
            raise MCPAppChannelError(
                "invalid_jsonrpc", "MCP App message has an invalid JSON-RPC envelope"
            )
        if not _constant_time_text(payload.get("channelId"), self.channel_id):
            raise MCPAppChannelError(
                "channel_mismatch", "MCP App message channel does not match its frame"
            )
        if not _constant_time_text(payload.get("nonce"), self.nonce):
            raise MCPAppChannelError(
                "nonce_mismatch", "MCP App message nonce does not match its frame"
            )
        request_id = payload.get("id")
        if isinstance(request_id, bool) or not isinstance(request_id, (str, int)):
            raise MCPAppChannelError(
                "invalid_request_id", "MCP App request id must be a string or integer"
            )
        if isinstance(request_id, str) and (not request_id or len(request_id) > 128):
            raise MCPAppChannelError(
                "invalid_request_id", "MCP App string request id is not bounded"
            )
        if request_id in self._seen:
            raise MCPAppChannelError(
                "replayed_request", "MCP App request id has already been used"
            )
        method = payload.get("method")
        if not isinstance(method, str) or method not in MCP_APP_ALLOWED_METHODS:
            raise MCPAppChannelError(
                "method_denied", "MCP App method is not allowed by host policy"
            )
        params = payload.get("params", {})
        if not isinstance(params, dict):
            raise MCPAppChannelError(
                "invalid_params", "MCP App JSON-RPC params must be an object"
            )
        _validate_method_params(method, params)
        if len(self._outstanding) >= self._limits.max_outstanding_app_requests:
            raise MCPAppChannelError(
                "outstanding_limit",
                "MCP App outstanding request limit reached",
            )
        self._seen.add(request_id)
        self._outstanding.add(request_id)
        return MCPAppBridgeRequest(
            request_id=request_id,
            method=method,
            params=params,
            binding=self.binding,
        )

    def _audit(
        self,
        code: str,
        *,
        method: str | None = None,
        request_id: str | int | None = None,
    ) -> None:
        if self._record_event is None:
            return
        self._record_event(
            MCPAppSecurityEvent(
                instance_id=self.instance_id,
                server_id=self.binding.server_id,
                resource_sha256=self.binding.resource_sha256,
                code=code,
                method=method,
                request_id_digest=_request_id_digest(request_id),
            )
        )


def _constant_time_text(value: object, expected: str) -> bool:
    return isinstance(value, str) and hmac.compare_digest(value, expected)


def _request_id_digest(request_id: str | int | None) -> str | None:
    if request_id is None:
        return None
    typed_value = f"{type(request_id).__name__}:{request_id}".encode()
    return hashlib.sha256(typed_value).hexdigest()


def _json_depth(value: object, *, depth: int = 0) -> int:
    if depth > 32:
        return depth
    if isinstance(value, Mapping):
        return max(
            (_json_depth(item, depth=depth + 1) for item in value.values()),
            default=depth,
        )
    if isinstance(value, list):
        return max(
            (_json_depth(item, depth=depth + 1) for item in value), default=depth
        )
    return depth


def _validate_method_params(method: str, params: Mapping[str, object]) -> None:
    if method == "ui/ready" and params:
        raise MCPAppChannelError(
            "invalid_params", "MCP App ready request does not accept parameters"
        )
    if method != "ui/userChoice":
        return
    choice_id = params.get("choiceId")
    if set(params) != {"choiceId"} or not isinstance(choice_id, str):
        raise MCPAppChannelError(
            "invalid_params", "MCP App user choice requires one typed choiceId"
        )
    if not choice_id or len(choice_id.encode("utf-8")) > 256:
        raise MCPAppChannelError(
            "invalid_params", "MCP App choiceId is empty or exceeds its byte limit"
        )

File: web/src/entities/mcp-app/bridge.ts
import {
  MCP_APP_MAX_POST_MESSAGE_BYTES,
  MCP_APP_MAX_OUTSTANDING_REQUESTS,
} from "./model";
import type { McpAppFrameDescriptor } from "./model";

const allowedMethods = new Set(["ui/ready", "ui/userChoice"] as const);

export type McpAppAllowedMethod = "ui/ready" | "ui/userChoice";

export interface McpAppChannelEnvelope {
  jsonrpc: "2.0";
  id: string | number;
  method: McpAppAllowedMethod;
  params: Readonly<Record<string, unknown>>;
  channelId: string;
  nonce: string;
}

export interface McpAppChannelRequest {
  envelope: McpAppChannelEnvelope;
  serialized: string;
  byteLength: number;
}

export type McpAppChannelDecision =
  | { accepted: true; request: McpAppChannelRequest }
  | {
      accepted: false;
      code:
        | "source_mismatch"
        | "invalid_jsonrpc"
        | "payload_limit"
        | "channel_mismatch"
        | "nonce_mismatch"
        | "invalid_request_id"
        | "method_denied"
        | "invalid_params";
    };

export function inspectMcpAppMessage(
  event: Pick<MessageEvent<unknown>, "data" | "source">,
  expectedSource: Window,
  descriptor: McpAppFrameDescriptor,
): McpAppChannelDecision {
  if (event.source !== expectedSource) {
    return { accepted: false, code: "source_mismatch" };
  }
  if (!isRecord(event.data) || jsonDepth(event.data) > 32) {
    return { accepted: false, code: "invalid_jsonrpc" };
  }
  const allowedKeys = new Set([
    "jsonrpc",
    "id",
    "method",
    "params",
    "channelId",
    "nonce",
  ]);
  if (
    Object.keys(event.data).some((key) => !allowedKeys.has(key)) ||
    event.data.jsonrpc !== "2.0"
  ) {
    return { accepted: false, code: "invalid_jsonrpc" };
  }
  if (event.data.channelId !== descriptor.channel_id) {
    return { accepted: false, code: "channel_mismatch" };
  }
  if (event.data.nonce !== descriptor.nonce) {
    return { accepted: false, code: "nonce_mismatch" };
  }
  const requestId = event.data.id;
  if (
    (typeof requestId !== "string" && typeof requestId !== "number") ||
    typeof requestId === "boolean" ||
    (typeof requestId === "string" && (requestId.length === 0 || requestId.length > 128))
  ) {
    return { accepted: false, code: "invalid_request_id" };
  }
  const method = event.data.method;
  if (typeof method !== "string" || !isAllowedMethod(method)) {
    return { accepted: false, code: "method_denied" };
  }
  const params = event.data.params ?? {};
  if (!isRecord(params) || !validParams(method, params)) {
    return { accepted: false, code: "invalid_params" };
  }
  const envelope: McpAppChannelEnvelope = {
    channelId: descriptor.channel_id,
    id: requestId,
    jsonrpc: "2.0",
    method,
    nonce: descriptor.nonce,
    params,
  };
  let serialized: string;
  try {
    serialized = JSON.stringify(envelope);
  } catch {
    return { accepted: false, code: "invalid_jsonrpc" };
  }
  const byteLength = new TextEncoder().encode(serialized).byteLength;
  if (byteLength === 0 || byteLength > MCP_APP_MAX_POST_MESSAGE_BYTES) {
    return { accepted: false, code: "payload_limit" };
  }
  return {
    accepted: true,
    request: { byteLength, envelope, serialized },
  };
}

export class McpAppRequestWindow {
  readonly limit = MCP_APP_MAX_OUTSTANDING_REQUESTS;
  readonly #pending = new Map<string, AbortController>();
  readonly #seen = new Set<string>();

  get size(): number {
    return this.#pending.size;
  }

  admit(requestId: string | number): AbortController | null {
    const key = typedRequestId(requestId);
    if (
      this.#seen.has(key) ||
      this.#seen.size >= this.limit * 256 ||
      this.#pending.size >= this.limit
    ) {
      return null;
    }
    const controller = new AbortController();
    this.#seen.add(key);
    this.#pending.set(key, controller);
    return controller;
  }

  complete(requestId: string | number): void {
    this.#pending.delete(typedRequestId(requestId));
  }

  cancelAll(): void {
    for (const controller of this.#pending.values()) controller.abort();
    this.#pending.clear();
    this.#seen.clear();
  }
}

function typedRequestId(requestId: string | number): string {
  return `${typeof requestId}:${String(requestId)}`;
}

function isAllowedMethod(method: string): method is McpAppAllowedMethod {
  return allowedMethods.has(method as McpAppAllowedMethod);
}

function validParams(
  method: McpAppAllowedMethod,
  params: Readonly<Record<string, unknown>>,
): boolean {
  if (method === "ui/ready") return Object.keys(params).length === 0;
  const choiceId = params.choiceId;
  return (
    Object.keys(params).length === 1 &&
    typeof choiceId === "string" &&
    choiceId.length > 0 &&
    new TextEncoder().encode(choiceId).byteLength <= 256
  );
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function jsonDepth(value: unknown, depth = 0): number {
  if (depth > 32) return depth;
  if (Array.isArray(value)) {
    return value.reduce(
      (maximum, item) => Math.max(maximum, jsonDepth(item, depth + 1)),
      depth,
    );
  }
  if (!isRecord(value)) return depth;
  return Object.values(value).reduce<number>(
    (maximum, item) => Math.max(maximum, jsonDepth(item, depth + 1)),
    depth,
  );
}
